[ENG-7289] Fix Search Index Discrepancy in Collection Facets

antkryt opened the pull request:
<!-- Before submit your Pull Request, make sure you picked
     the right branch:

     - For hotfixes, select "master" as the target branch
     - For new features, select "develop" as the target branch
     - For release feature fixes, select the relevant release branch (release/X.Y.Z) as the target branch -->

## Purpose

Update SHARE index when collection submission is created

## Changes

- add update_share() call
- minor collection submission transitions refactor to remove redundant transitions

## QA Notes

This PR won't fix collection submission indexes that were not created, but will prevent such situation in the future 

## Documentation

<!-- Does any internal or external documentation need to be updated?
     - If the API was versioned, update the developer.osf.io changelog.
     - If changes were made to the API, link the developer.osf.io PR here.
-->

## Side Effects

<!-- Any possible side effects? -->

## Ticket

https://openscience.atlassian.net/browse/ENG-7289

brianjgeiger reviewed (changes requested):
Two things:

1. The ticket specifically called for fixing the existing collection submissions that weren't indexed. It's great that the problem is fixed going forward, but we'll also need a management command that reindexes all of the existing collection submissions in SHARE.
2. The changes to the state machine seem out of scope for this ticket. Are they necessary for your changes to work? This is going to involve a complete regression test of collections and collections moderation in order to pass, which is more work than I was expecting to put QA through for this ticket.

antkryt commented:
1. We already have it, see osf/management/commands/sync_collection_provider_indices.py. You can use it as follows:
`docker compose run --rm web python3 manage.py sync_collection_provider_indices --collection_provider_ids studyswap flubber`
Not sure how many collection providers we have on prod, but I wouldn’t recommend updating all of them at once. Each collection submission kicks off a Celery task that hits SHARE, and doing that for everything at once may basically cause a DOS attack
2. FYI, with the current implementation when we try to add 1 submission, we make 5 identical SHARE requests. So I tried to implement the `update_share` call as part of the workflow to make only one SHARE request, but it was not possible. That’s where those changes come from, and I decided to leave it since they don't change anything just make it cleaner. In any case, I removed it to prevent additional testing

Additionally, I've added test that ensures that update_share is called when we create collection submission
